torch/csrc/python_dimname.cpp

```diff
@@ -0,0 +1,83 @@
+#ifdef NAMEDTENSOR_ENABLED
+#include <torch/csrc/python_dimname.h>
+#include <torch/csrc/Exceptions.h>
+#include <torch/csrc/utils/python_strings.h>
+#include <c10/util/flat_hash_map.h>
+
+namespace torch {
+
+struct InternedStringsTable {
+  InternedStringsTable() = default;
+  ~InternedStringsTable();
+  InternedStringsTable(const InternedStringsTable &) = delete;
+  InternedStringsTable& operator =(InternedStringsTable const&) = delete;
+  InternedStringsTable(InternedStringsTable&&) = delete;
+  InternedStringsTable& operator=(InternedStringsTable&&) = delete;
+
+  at::optional<at::Dimname> lookup(PyObject* obj);
+  // Precondition: obj is an interned python string.
+  void addMapping(PyObject* obj, at::Dimname dimname);
+ private:
+  ska::flat_hash_map<PyObject*,at::Dimname> py_interned_string_to_dimname_;
+};
+
+InternedStringsTable kPyInternedStringToDimname;
+
+InternedStringsTable::~InternedStringsTable() {
+  for (auto it = py_interned_string_to_dimname_.begin();
+      it != py_interned_string_to_dimname_.end(); ++it) {
+    // See Note [References to python interned strings]
+    Py_DECREF(it->first);
+  }
+}
+
+at::optional<at::Dimname> InternedStringsTable::lookup(PyObject* obj) {
+  auto it = py_interned_string_to_dimname_.find(obj);
+  if (it == py_interned_string_to_dimname_.end()) {
+    return at::nullopt;
+  }
+  return it->second;
+}
+
+
+void InternedStringsTable::addMapping(PyObject* obj, at::Dimname dimname) {
+  // Note [References to python interned strings]
+  // If a Python interned string has no references to it, then it gets
+  // deallocated, invalidating this mapping. Let's immortalize the string by
+  // holding a refcount to it and releasing it in the destructor
+  Py_INCREF(obj);
+  py_interned_string_to_dimname_.emplace(obj, dimname);
+}
+
+} // namespace torch
+
+at::Dimname THPDimname_parse(PyObject* obj) {
+  if (obj == Py_None) {
+    return at::Dimname::wildcard();
+  }
+
+  if (!THPUtils_checkString(obj)) {
+    throw torch::TypeError("expected None or string for Dimname but got %s", Py_TYPE(obj)->tp_name);
+  }
+
+  if (!THPUtils_isInterned(obj)) {
+    // internStringInPlace decrefs obj and increfs the result. Because we're
+    // not actually returning the result to the user, we need to undo these.
+    // See https://docs.python.org/3/c-api/unicode.html#c.PyUnicode_InternInPlace
+    Py_INCREF(obj);
+    THPUtils_internStringInPlace(&obj);
+    Py_DECREF(obj);
```

Comment:
is there somehow we can test this interning system / that it properly allows freeing the memory for the dim names etc? I'd imagine it's not a very common case that a dim name goes completely out of scope and stops being used, and dim names are small, so would be easy to not notice if this doesn't actually work correctly.

Reply:
I don't have a good idea for how to test this. It's not easy to use mocks because we have global state running around in a number of places. I'll go and think about this a bit more but I am taking suggestions

Reply:
Okay, might be possible to test this in Python: https://stackoverflow.com/questions/510406/is-there-a-way-to-get-the-current-ref-count-of-an-object-in-python . Going to see what I can do here.

Reply:
It looks like we can count total reference counts using sys.gettotalrefcount()

https://pythonextensionpatterns.readthedocs.io/en/latest/debugging/leak_newrefs_vg.html#observing-the-total-reference-counts

We could verify it goes up/down by an expected amount if we create random dimname strings and then let them go out of scope.

Reply:
Yup, I added some tests for the refcounting behavior.